qt/scientific_interfaces/Indirect/IndirectFitOutput.cpp

```diff
@@ -0,0 +1,213 @@
+#include "IndirectFitOutput.h"
+
+#include "MantidAPI/TableRow.h"
+
+using namespace Mantid::API;
+
+namespace {
+using namespace MantidQt::CustomInterfaces::IDA;
+
+struct TableRowExtractor {
+  explicit TableRowExtractor(ITableWorkspace_sptr table)
+      : m_table(table), m_columns(m_table->getColumnNames()) {
+    m_chiIndex = std::find(m_columns.begin(), m_columns.end(), "Chi_squared") -
+                 m_columns.begin();
+  }
+
+  std::unordered_map<std::string, ParameterValue>
+  operator()(std::size_t index) {
+    TableRow row = m_table->getRow(index);
+    std::unordered_map<std::string, ParameterValue> parameters;
+
+    for (auto i = 1u; i < m_chiIndex; i += 2) {
+      const auto &columnName = m_columns[i];
+      parameters[columnName] = ParameterValue(row.Double(i), row.Double(i + 1));
+    }
+    return parameters;
+  }
+
+private:
+  ITableWorkspace_sptr m_table;
+  const std::vector<std::string> m_columns;
+  std::size_t m_chiIndex;
+};
+
+template <typename Map, typename Key>
+typename Map::mapped_type &extractOrAddDefault(Map &map, const Key &key) {
+  auto values = map.find(key);
+  if (values != map.end())
+    return values->second;
+  return map[key] = typename Map::mapped_type();
+}
+
+template <typename F>
+void applyEnumeratedData(const F &functor, const FitDataIterator &fitDataBegin,
+                         const FitDataIterator &fitDataEnd) {
+  std::size_t start = 0;
+  for (auto it = fitDataBegin; it < fitDataEnd; ++it)
+    start = (*it)->applyEnumeratedSpectra(functor(it->get()), start);
+}
+
+void extractParametersFromTable(
+    ITableWorkspace_sptr tableWs, const FitDataIterator &fitDataBegin,
+    const FitDataIterator &fitDataEnd,
+    std::unordered_map<IndirectFitData const *, ParameterValues> &parameters) {
+  TableRowExtractor extractRowFromTable(tableWs);
+  auto extract = [&](IndirectFitData const *inputData) {
+    auto &values = extractOrAddDefault(parameters, inputData);
+    return [&](std::size_t index, std::size_t spectrum) {
+      values[spectrum] = extractRowFromTable(index);
+    };
+  };
+  applyEnumeratedData(extract, fitDataBegin, fitDataEnd);
+}
+
+template <typename Map2D, typename T, typename U, typename Value>
+Value findOr(const Map2D &map, const T &firstKey, const U &secondKey,
```

Comment:
I struggle to understand what this function does from the name (you have another function called `extractOrAddDefault` so maybe this one could be `findOrAddDefault`)...

Just out of interest, is there a way to make this recursive so it would apply to arbitrarily nested maps?

Reply:
This method doesn't add to the map, unlike extractOrAddDefault, so probably shouldn't be called findOrAddDefault. It was modelled after boost::optional's get_value_or -- it may however still be best to rename.

Reply:
Ok fair enough. If it makes sense to people familiar with boost then this isn't an issue